mls-rs-uniffi/src/lib.rs

```diff
@@ -669,3 +669,127 @@ impl Group {
+            }
+        }
+
+        impl GroupStateStorage for CustomGroupStateStorage {
+            fn state(&self, group_id: Vec<u8>) -> Result<Option<Vec<u8>>, FFICallbackError> {
+                let groups = self.lock();
+                Ok(groups.get(&group_id).map(|group| group.state.clone()))
+            }
+
+            fn epoch(
+                &self,
+                group_id: Vec<u8>,
+                epoch_id: u64,
+            ) -> Result<Option<Vec<u8>>, FFICallbackError> {
+                let groups = self.lock();
+                match groups.get(&group_id) {
+                    Some(group) => {
+                        let data = group
+                            .epoch_data
+                            .iter()
+                            .find_map(|EpochRecord { id, data }| {
+                                (*id == epoch_id).then_some(data.clone())
+                            });
+                        Ok(data)
+                    }
+                    None => Ok(None),
+                }
+            }
+
+            fn write(
+                &self,
+                state: GroupState,
+                epoch_inserts: Vec<EpochRecord>,
+                epoch_updates: Vec<EpochRecord>,
+            ) -> Result<(), FFICallbackError> {
+                let mut groups = self.lock();
+
+                let group = groups.entry(state.id).or_insert(GroupStateData {
+                    state: state.data,
+                    epoch_data: vec![],
+                });
+                for insert in epoch_inserts {
+                    group.epoch_data.push(insert);
+                }
+
+                for update in epoch_updates {
+                    for epoch in group.epoch_data.iter_mut() {
+                        if epoch.id == update.id {
+                            epoch.data = update.data;
+                            break;
+                        }
+                    }
+                }
+
+                Ok(())
+            }
+
+            fn max_epoch_id(&self, group_id: Vec<u8>) -> Result<Option<u64>, FFICallbackError> {
+                let groups = self.lock();
+                Ok(groups
+                    .get(&group_id)
+                    .and_then(|GroupStateData { epoch_data, .. }| epoch_data.last())
+                    .map(|last| last.id))
+            }
+        }
+
+        let group_state_storage = Arc::new(CustomGroupStateStorage {
+            groups: Mutex::new(HashMap::new()),
+        });
+        let client_config = ClientConfig {
+            group_state_storage: group_state_storage.clone(),
+        };
+
+        let alice_keypair = generate_signature_keypair(CipherSuite::Curve25519Aes128)?;
+        let alice = Client::new(b"alice".to_vec(), alice_keypair, client_config.clone());
+
+        let bob_keypair = generate_signature_keypair(CipherSuite::Curve25519Aes128)?;
+        let bob = Client::new(b"bob".to_vec(), bob_keypair, client_config.clone());
```

Comment:
`client_config` contains an `Arc` to the storage, so cloning it means it is shared and Alice and Bob will clobber each other's data.

Reply:
> `client_config` contains an `Arc` to the storage, so cloning it means it is shared and Alice and Bob will clobber each other's data.

Oh, good point! I wrote the Rust code here by following the [Python test](https://github.com/awslabs/mls-rs/blob/main/mls-rs-uniffi/tests/simple_scenario_sync.py), so perhaps that has the same problem? Cc @tomleavy.